src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioDiagnosticListSuppressionStateService.cs

```diff
@@ -0,0 +1,253 @@
+            _tableControl = errorList?.TableControl;
+
+            ClearState();
+            InitializeFromTableControlIfNeeded();
+        }
+
+        private int SelectedItems => _selectedActiveItems + _selectedSuppressedItems + _selectedNonSuppressionStateItems;
+
+        // If we can suppress either in source or in suppression file, we enable suppress context menu.
+        public bool CanSuppressSelectedEntries => CanSuppressSelectedEntriesInSource || CanSuppressSelectedEntriesInSuppressionFiles;
+
+        // If only suppressed items are selected, we enable remove suppressions.
+        public bool CanRemoveSuppressionsSelectedEntries => _selectedActiveItems == 0 && _selectedSuppressedItems > 0;
+
+        // If only Roslyn active items are selected, we enable suppress in source.
+        public bool CanSuppressSelectedEntriesInSource => _selectedActiveItems > 0 &&
+            _selectedSuppressedItems == 0 &&
+            _selectedRoslynItems == _selectedActiveItems;
+
+        // If only active items are selected, and there is at least one Roslyn item, we enable suppress in suppression file.
+        // Also, compiler diagnostics cannot be suppressed in suppression file, so there must be at least one non-compiler item.
+        public bool CanSuppressSelectedEntriesInSuppressionFiles => _selectedActiveItems > 0 &&
+            _selectedSuppressedItems == 0 &&
+            (_selectedRoslynItems - _selectedCompilerDiagnosticItems) > 0;
+
+        private void ClearState()
+        {
+            _selectedActiveItems = 0;
+            _selectedSuppressedItems = 0;
+            _selectedRoslynItems = 0;
+            _selectedCompilerDiagnosticItems = 0;
+            _selectedNonSuppressionStateItems = 0;
+        }
+
+        private void InitializeFromTableControlIfNeeded()
+        {
+            if (_tableControl == null)
+            {
+                return;
+            }
+
+            if (SelectedItems == _tableControl.SelectedEntries.Count())
+            {
+                // We already have up-to-date state data, so don't need to re-compute.
+                return;
+            }
+
+            ClearState();
+            if (ProcessEntries(_tableControl.SelectedEntries, added: true))
+            {
+                UpdateQueryStatus();
+            }
+        }
+
+        public void ProcessSelectionChanged(TableSelectionChangedEventArgs e)
+        {
+            var hasAddedSuppressionStateEntry = ProcessEntries(e.AddedEntries, added: true);
+            var hasRemovedSuppressionStateEntry = ProcessEntries(e.RemovedEntries, added: false);
+
+            // If any entry that supports suppression state was ever involved, update query status since each item in the error list
+            // can have different context menu.
+            if (hasAddedSuppressionStateEntry || hasRemovedSuppressionStateEntry)
+            {
+                UpdateQueryStatus();
```

Comment:
it only processes newly selected entries, so should be fast. again I took it from current FXCop event processor, but probably needs more perf testing?